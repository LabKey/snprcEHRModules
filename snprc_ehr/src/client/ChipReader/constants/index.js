export default {
    debug: false,
    notSupportedMessage: `Sorry, Serial.API is not supported on this device, make sure you're 
    running Chrome 78 or later and have enabled the #enable-experimental-web-platform-features flag in
    chrome://flags`,
    timeOutErrorMessage: 'Chip reader is not responding. Ensure you are connected to the correct port, and the reader is turned on.',
    validSerialOptions: [ 'baudrate', 'databits', 'stopbits', 'parity', 'buffersize', 'rtscts', 'xon', 'xoff', 'xany' ],
    defaultSerialOptions: {
        baudrate: 19200,
        databits: 8,
        stopbits: 1,
        buffersize: 255,
        parity: 'none',
        rctscts: false,
        xon: true,
        xoff: true,
        xany: false
    }
}

// https://wicg.github.io/serial/#dom-serial
// parity is an enum:   "none",  "even",  "odd",  "mark",  "space"
// baudrate member: One of 115200, 57600, 38400, 19200, 9600, 4800, 2400, 1800, 1200, 600, 300, 200, 150, 134, 110, 75, or 50.
// databits member: One of 8, 7, 6, or 5.
// stopbits member: One of 1 or 2.
// rctscts, xon, xoff, xany: boolean