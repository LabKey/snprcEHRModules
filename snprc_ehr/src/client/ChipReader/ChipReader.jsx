/* eslint-disable no-alert */

import React from 'react'
import { LoadingSpinner } from '../Shared/components/LoadingSpinner'
import './styles/chipReader.scss'
import ChipReaderState from './constants/chipReaderState'
import constants from './constants/index'
import InfoPanel from '../Shared/components/InfoPanel'
import ChipDataPanel from './components/ChipDataPanel'
import SummaryGridPanel from './components/SummaryGridPanel'
import fetchAnimalId from './api/fetchAnimalId'
import { close } from './services/serialService'

export default class ChipReader extends React.Component {
    state = new ChipReaderState();
    notSupportedMessage = constants.notSupportedMessage
    debug = constants.debug;
    isSerialSupported = ('serial' in navigator) || this.props.debug
    previousChipId = undefined
    componentDidMount() {
        window.addEventListener('beforeunload', this.beforeunload.bind(this))

        this.setState(prevState => (
            {
                ...prevState,
                serialOptions: constants.defaultSerialOptions,
                isLoading: false
            }
        ))
    }
    componentWillUnmount() {
        window.removeEventListener('beforeunload', this.beforeunload.bind(this))
    }
    beforeunload(e) {
        if (this.state.connection) {
            close(this.state.connection)
            e.preventDefault()
            e.returnValue = true
        }
    }
    handleSetConnection = connection => {
        this.setState(prevState => (
            {
                ...prevState,
                connection
            }
        ))
    }
    handleDataChange = async value => {
        const data = value || { chipId: undefined, animalId: undefined, temperature: undefined }

        if (!data.chipId || data.chipId === this.previousChipId) return

        // console.log(`chipId = ${data.chipId}`)
        this.previousChipId = data.chipId

        if (data.chipId && !data.animalId) {
            data.animalId = await fetchAnimalId(data.chipId).catch(error => {
                this.setState(prevState => (
                    {
                        ...prevState,
                        errorMessage: error.message
                    }
                ))
            })
        }

        this.setState(prevState => ({
            ...prevState,
            chipData: data,
            ...(data.animalId && { errorMessage: undefined }), // clear error message
            ...(data.animalId && {
                summaryData: [
                    ...prevState.summaryData,
                    { ...data }
                ]
            })
        }))
    }
    handleErrorMessage = error => {
        this.setState(prevState => (
            {
                ...prevState,
                errorMessage: error
            }
        ))
    }
    render() {
        // allow debug mode to be triggered for running test suite
        this.debug = this.props.debug !== undefined ? this.props.debug : constants.debug

        const { isLoading } = this.state

        if (isLoading) {
            return (
              <LoadingSpinner msg="Loading app..." />
            )
        }

        return (
          <div>
            { this.isSerialSupported
                    && (
                    <div className="split-panel">
                      <div className="parent-panel">
                        <div className="panel-heading">
                          <p>Current Animal</p>
                        </div>
                        <div className="wizard-panel">
                          <ChipDataPanel
                            chipData={ this.state.chipData }
                            connection={ this.state.connection }
                            errorMessage={ this.state.errorMessage }
                            handleDataChange={ this.handleDataChange }
                            handleErrorMessage={ this.handleErrorMessage }
                            handleSetConnection={ this.handleSetConnection }
                            serialOptions={ this.state.serialOptions }
                          />
                        </div>

                      </div>

                      <div className="right-panel">
                        <div>
                          <div className="panel-heading">
                            <p>Animals Seen</p>
                          </div>
                          <div className="wizard-right">
                            <SummaryGridPanel
                              summaryData={ this.state.summaryData }
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                    ) }
            <InfoPanel
              errorMessages={ (!this.isSerialSupported
                        && [{ propTest: true, colName: this.notSupportedMessage }])
                        || (this.state.errorMessage && [{ propTest: true, colName: this.state.errorMessage }]) }
            />
          </div>
        )
    }
}
