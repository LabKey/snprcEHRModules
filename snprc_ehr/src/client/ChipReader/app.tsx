import React from 'react'
import ReactDOM from 'react-dom'

import ChipReader from './ChipReader.jsx'

// Need to wait for container element to be available in labkey wrapper before render
window.addEventListener('DOMContentLoaded', () => {
    ReactDOM.render(<ChipReader />, document.getElementById('app'))
})
