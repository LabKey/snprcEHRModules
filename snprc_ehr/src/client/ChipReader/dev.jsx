import React from 'react'
import ReactDOM from 'react-dom'
import { AppContainer } from 'react-hot-loader'

import ChipReader from './ChipReader'

const render = () => {
    ReactDOM.render(
      <AppContainer>
        <ChipReader />
      </AppContainer>,
        document.getElementById('app')
    )
}

if (module.hot) {
    module.hot.accept()
}

render()
