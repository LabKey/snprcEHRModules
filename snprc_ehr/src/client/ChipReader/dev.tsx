import React from 'react'
import ReactDOM from 'react-dom'
import { AppContainer } from 'react-hot-loader'

import { ChipReader } from './ChipReader.jsx'

const render = () => {
    ReactDOM.render(
      <AppContainer>
        <ChipReader />
      </AppContainer>,
        document.getElementById('app')
    )
}

declare const module: any;

if (module.hot) {
    module.hot.accept()
}

render()
