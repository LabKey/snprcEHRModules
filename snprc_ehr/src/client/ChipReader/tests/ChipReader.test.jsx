/* eslint-disable no-unused-vars */

import React from 'react'
import { shallow } from 'enzyme'
import ChipReader from '../ChipReader'

require("setimmediate")

jest.mock('../../Shared/api/api')

beforeAll(() => {
    // global.fetch = jest.fn();
})

let wrapper

beforeEach(() => {
    wrapper = shallow(<ChipReader debug />)
})

afterEach(() => {
    wrapper.unmount()
})

function flushPromises() {
    return new Promise(resolve => setImmediate(resolve))
}

describe('ChipReader tests', () => {
    test('Should render the landing page with no serial support error.', async () => {
        wrapper = shallow(<ChipReader />)
        await flushPromises()
        expect(wrapper).toMatchSnapshot()
    })

    test('Should render the landing page without error message.', async () => {
        await flushPromises()
        expect(wrapper).toMatchSnapshot()

        // spinner should be gone
        expect(wrapper.find('LoadingSpinner').exists()).toBeFalsy()
        // ChipDataPanel and SummaryGridPanel panels should be present
        expect(wrapper.find('ChipDataPanel').exists()).toBeTruthy()
        expect(wrapper.find('SummaryGridPanel').exists()).toBeTruthy()
    })

    test('Should render a loading spinner before rendering the landing page.', async () => {
        wrapper.unmount()
        wrapper = shallow(<ChipReader />, { disableLifecycleMethods: true })
        await flushPromises()
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('LoadingSpinner').exists()).toBeTruthy()
    })
})
