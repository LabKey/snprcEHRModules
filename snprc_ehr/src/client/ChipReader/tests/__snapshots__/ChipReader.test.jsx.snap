// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipReader tests Should render a loading spinner before rendering the landing page. 1`] = `
<LoadingSpinner_LoadingSpinner
  msg="Loading app..."
  wrapperClassName=""
/>
`;

exports[`ChipReader tests Should render the landing page with no serial support error. 1`] = `
<div>
  <InfoPanel
    errorMessages={
      Array [
        Object {
          "colName": "Sorry, Serial.API is not supported on this device, make sure you're 
    running Chrome 78 or later and have enabled the #enable-experimental-web-platform-features flag in
    chrome://flags",
          "propTest": true,
        },
      ]
    }
  />
</div>
`;

exports[`ChipReader tests Should render the landing page without error message. 1`] = `
<div>
  <div
    className="split-panel"
  >
    <div
      className="parent-panel"
    >
      <div
        className="panel-heading"
      >
        <p>
          Current Animal
        </p>
      </div>
      <div
        className="wizard-panel"
      >
        <ChipDataPanel
          chipData={
            Object {
              "animalId": undefined,
              "chipId": undefined,
              "temperature": undefined,
              "time": undefined,
            }
          }
          handleDataChange={[Function]}
          handleErrorMessage={[Function]}
          handleSetConnection={[Function]}
          serialOptions={
            Object {
              "baudrate": 19200,
              "buffersize": 255,
              "databits": 8,
              "parity": "none",
              "rctscts": false,
              "stopbits": 1,
              "xany": false,
              "xoff": true,
              "xon": true,
            }
          }
        />
      </div>
    </div>
    <div
      className="right-panel"
    >
      <div>
        <div
          className="panel-heading"
        >
          <p>
            Animals Seen
          </p>
        </div>
        <div
          className="wizard-right"
        >
          <SummaryGridPanel
            summaryData={Array []}
          />
        </div>
      </div>
    </div>
  </div>
  <InfoPanel />
</div>
`;
